### PR TITLE
SAMZA-1051: Merge operator APIs to master

### PR DESCRIPTION
**Pull request by nickpan47**
Merged all changes from operator APIs and added interface stability annotation.

### CODE REVIEW

**Comment by nickpan47** [comment]
@vjagadish1989 @navina could you take a look and give a ship it on this one?

File: build.gradle [hunk 1]
```diff
@@ -370,6 +370,24 @@ project(":samza-yarn_$scalaVersion") {
   jar.dependsOn("lesscss")
 }
 
+project(":samza-operator") {
+  apply plugin: 'java'
+
```
Comment (vjagadish1989) [inline build.gradle]: Can we enable checkstyle back?

Comment (vjagadish1989) [inline build.gradle]: Can we re-enable checkstyle?  I believe the most recent patch had it enabled?

File: samza-api/src/main/java/org/apache/samza/annotation/InterfaceStability.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+/**
+ * Annotation to inform users of how much to rely on a particular package,
+ * class or method not changing over time. Currently the stability can be
+ * {@link Stable}, {@link Evolving} or {@link Unstable}. <br>
+ */
+@InterfaceStability.Evolving
+public class InterfaceStability {
```
Comment (vjagadish1989) [inline samza-api/src/main/java/org/apache/samza/annotation/InterfaceStability.java]: Big +1 for this class. 👍

File: samza-api/src/main/java/org/apache/samza/operators/Windows.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+  /**
+   * The public programming interface class for window function
+   *
+   * @param <M>  the type of input {@link Message}
+   * @param <WK>  the type of key to the {@link Window}
+   * @param <WV>  the type of output value in the {@link WindowOutput}
+   * @param <WM>  the type of message in the window output stream
+   */
+  public interface Window<M extends Message, WK, WV, WM extends WindowOutput<WK, WV>> {
```
Comment (vjagadish1989) [inline samza-api/src/main/java/org/apache/samza/operators/Windows.java]: probably need an Unstable annotation here?

File: settings.gradle [hunk 4]
```diff
@@ -21,7 +21,8 @@ include \
   'samza-elasticsearch',
   'samza-log4j',
   'samza-shell',
-  'samza-rest'
+  'samza-rest',
+  'samza-operator'
```
Comment (prateekm) [inline settings.gradle]: Minor: alphabetize :)

File: samza-api/src/main/java/org/apache/samza/operators/Windows.java (hunk 3 above)
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/operators/Windows.java]: Yes. missed this one! Thanks!

Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/operators/Windows.java]: @vjagadish1989 yes. It is logged as SAMZA-1052. I realize that the integration step would remove samza-operator module completely, which creates big conflicts w/ @prateekm 's refactoring effort. Hence, hold up on that.

File: build.gradle [hunk 5]
```diff
@@ -370,6 +370,24 @@ project(":samza-yarn_$scalaVersion") {
...
+project(":samza-operator") {
+  apply plugin: 'java'
+    
+  sourceCompatibility = 1.8
+    
+  dependencies {
+    compile project(':samza-api')
+    compile project(":samza-core_$scalaVersion")
+    compile "commons-collections:commons-collections:$commonsCollectionVersion"
+    compile "org.apache.commons:commons-lang3:$commonsLang3Version"
```
Comment (nickpan47) [inline build.gradle]: Good catch. I removed commons-lang3.

**Comment by nickpan47** [comment]
Committed.

**Pull request closed without merging**